Authentication claims mapping and claims transformation (#22456)
dotnet/AspNetCore.Docs: aspnetcore/security/authentication/claims.md

---

> title: Mapping claims, customizing claims and transforming claims in ASP.NET Core

Reviewer: ```suggestion
title: Mapping, customizing, and transforming claims in ASP.NET Core
```

Author: fixed

author: damienbod
description: Learn how to map claims, do claims transformations, customize claims.
monikerRange: '>= aspnetcore-3.1'
ms.author: riande
ms.custom: mvc
ms.date: 05/29/2021
no-loc: [Home, Privacy, Kestrel, appsettings.json, "ASP.NET Core Identity", cookie, Cookie, Blazor, "Blazor Server", "Blazor WebAssembly", "Identity", "Let's Encrypt", Razor, SignalR]
uid: security/authentication/claims
---

> # Mapping claims, customizing claims and transforming claims in ASP.NET Core

Reviewer: ```suggestion
# Mapping, customizing, and transforming claims in ASP.NET Core
```

Author: fixed

By [Damien Bowden](https://github.com/damienbod)

> Claims can be created from any user data or identity data which can be issued using a trusted identity provider or ASP.NET Core identity. A claim is a name value pair that represents what the subject is, not what the subject can do.

Reviewer: ```suggestion
Claims can be created from any user or identity data which can be issued using a trusted identity provider or ASP.NET Core identity. A claim is a name value pair that represents what the subject is, not what the subject can do.
```

Author: fixed

This article covers the following areas:

> * How to configure and map claims using an OpenID Connect client

Reviewer: ```suggestion
* How to configure and map claims using an [OpenID Connect](https://openid.net/connect/) client
```

Author: fixed

> * Set the name claim and the role claim

Reviewer: ```suggestion
* Set the name and role claim
```

Author: fixed

* Reset the claims namespaces

> * Customize, extend the claims using IClaimsTransformation

Reviewer: ```suggestion
* Customize, extend the claims using <xref:Microsoft.AspNetCore.Authentication.IClaimsTransformation>
```

That's the interface, Or do you want the method: <xref:Microsoft.AspNetCore.Authentication.IClaimsTransformation.TransformAsync%2A>

Author: I customized using the IClaimsTransformation interface which has only one method, like you mentioned, so I'm not sure which is better

maybe this would be better, yes

Customize, extend the claims using <xref:Microsoft.AspNetCore.Authentication.IClaimsTransformation.TransformAsync>.

Author: Updated to use the TransformAsync as well

## Mapping claims using Open ID Connect authentication

> The profile claims can be returned in the **id_token** which is returned after a successful authentication. The ASP.NET Core client application just needs to request the profile scope. When using the id_token for claims, no extra claims mapping is required.

Reviewer: ```suggestion
The profile claims can be returned in the `id_token` which is returned after a successful authentication. The ASP.NET Core client app just needs to request the profile scope. When using the `id_token` for claims, no extra claims mapping is required.
```

Author: fixed

```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddAuthentication(options =>
    {
        options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
        options.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
    })
   .AddCookie()
   .AddOpenIdConnect(options =>
   {
       options.SignInScheme = "Cookies";
       options.Authority = "-your-identity-provider-";
       options.RequireHttpsMetadata = true;
       options.ClientId = "-your-clientid-";
       options.ClientSecret = "-your-client-secret-from-user-secrets-or-keyvault";
       options.ResponseType = "code";
       options.UsePkce = true;
       options.Scope.Add("profile");
       options.SaveTokens = true;
   });
```

> A second way to get the user claims is to use the OpenID Connect User Info API. The ASP.NET Core client application uses the **GetClaimsFromUserInfoEndpoint** property to configure this. One important difference from the first settings, is that you must specify the claims you require using the **MapUniqueJsonKey** method, otherwise only the **name**, **given_name** and **email** standard claims will be available in the client application. The claims included in the id_token are mapped per default. This is the major difference to the first option. You must explicitly define some of the claims you require.

Reviewer: ```suggestion
Another way to get the user claims is to use the OpenID Connect User Info API. The ASP.NET Core client application uses the `GetClaimsFromUserInfoEndpoint` property to configure this. One important difference from the first settings, is that you must specify the claims you require using the `MapUniqueJsonKey` method, otherwise only the `name`, `given_name` and `email` standard claims will be available in the client application. The claims included in the `id_token` are mapped per default. This is the major difference to the first option. You must explicitly define some of the claims you require.
```

Author: fixed

   .AddCookie()
   .AddOpenIdConnect(options =>
   {
       options.SignInScheme = "Cookies";
       options.Authority = "-your-identity-provider-";
       options.RequireHttpsMetadata = true;
       options.ClientId = "-your-clientid-";
       options.ClientSecret = "-your-client-secret-from-user-secrets-or-keyvault";
       options.ResponseType = "code";
       options.UsePkce = true;
       options.Scope.Add("profile");
       options.SaveTokens = true;
       options.GetClaimsFromUserInfoEndpoint = true;
       options.ClaimActions.MapUniqueJsonKey("preferred_username", "preferred_username");
       options.ClaimActions.MapUniqueJsonKey("gender", "gender");
   }); 
```

## Name claim and role claim mapping

The **Name** claim and the **Role** claim are mapped to default properties in the ASP.NET Core HTTP context. Sometimes it is required to use different claims for the default properties, or the name claim and the role claim do not match the default values. The claims can be mapped using the **TokenValidationParameters** property and set to any claim as required. The values from the claims can be used directly in the HttpContext **User.Identity.Name** property and the roles.

> If the **User.Identity.Name** has no value or the roles are missing, please check the values in the returned claims and set the **NameClaimType** and the **RoleClaimType** values. The returned claims from the client authentication can be viewed in the HTTP context.

Reviewer: ```suggestion
If the `User.Identity.Name` has no value or the roles are missing, please check the values in the returned claims and set the `NameClaimType` and the `RoleClaimType` values. The returned claims from the client authentication can be viewed in the HTTP context.
```

Author: fixed

```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddAuthentication(options =>
    {
        options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
        options.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
    })
   .AddCookie()
   .AddOpenIdConnect(options =>
   {
       // other options...
       options.TokenValidationParameters = new TokenValidationParameters
       {
         NameClaimType = "email", 
         // RoleClaimType = "role"
       };
   });
```

## Claims namespaces, default namespaces

> ASP.NET Core adds default namespaces to some known claims which might not be required in your application. You can disable this namespace adding, if you would like to use the claims exactly like the Open ID Connect server created. 

Reviewer: ```suggestion
ASP.NET Core adds default namespaces to some known claims which might not be required in your app. You can disable this namespace adding if you would like to use the claims exactly like the Open ID Connect server created. 
```

Author: fixed

```csharp
public void Configure(IApplicationBuilder app)
{
    JwtSecurityTokenHandler.DefaultInboundClaimTypeMap.Clear();
```

> ## Extending or adding custom claims using IClaimsTransformation

Reviewer: ```suggestion
## Extend or add custom claims using `IClaimsTransformation`
```

Author: fixed

> The **IClaimsTransformation** interface can be used to add extra claims to the **ClaimsPrincipal** class. The interface requires a single method **TransformAsync**. This method might get called multiple times. Only add a new claim if it does not already exist in the **ClaimsPrincipal**. A **ClaimsIdentity** is created to add the new claims and this can be added to the **ClaimsPrincipal**.

Reviewer: ```suggestion
The `IClaimsTransformation` interface can be used to add extra claims to the `ClaimsPrincipal` class. The interface requires a single method `TransformAsync`. This method might get called multiple times. Only add a new claim if it does not already exist in the `ClaimsPrincipal`. A `ClaimsIdentity` is created to add the new claims and this can be added to the `ClaimsPrincipal`.
```

Author: fixed

```csharp
public class MyClaimsTransformation : IClaimsTransformation
{
    public Task<ClaimsPrincipal> TransformAsync(ClaimsPrincipal principal)
    {
       ClaimsIdentity claimsIdentity = new ClaimsIdentity();
       var claimType = "myNewClaim";
       if (!principal.HasClaim(claim => claim.Type == claimType))
       {		   
          claimsIdentity.AddClaim(new Claim(claimType, "myClaimValue"));
       }

       principal.AddIdentity(claimsIdentity);
       return Task.FromResult(principal);
    }
}
```

> The **IClaimsTransformation** interface and the **MyClaimsTransformation** class can be added in the ConfigureServices method as a service.

Reviewer: ```suggestion
The `IClaimsTransformation` interface and the `MyClaimsTransformation` class can be added in the ConfigureServices method as a service.
```

Author: fixed

```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddTransient<IClaimsTransformation, MyClaimsTransformation>();
```

> ## Extending or adding custom claims in ASP.NET Core Identity

Reviewer: ```suggestion
## Extend or add custom claims in ASP.NET Core Identity
```

Author: fixed

> Please refer to this doc:

Reviewer: ```suggestion
Refer to the following document:
```

Author: fixed

[Add claims to Identity using IUserClaimsPrincipalFactory](xref:security/authentication/add-user-data#add-claims-to-identity-using-iuserclaimsprincipalfactory)

## Map claims from external identity providers

> Please refer to this doc:

Reviewer: ```suggestion
Refer to the following document:
```

Author: fixed
